src/codegen/generators/data_type_generator.rs

```diff
@@ -105,135 +21,153 @@ pub fn generate_data_type_stubs<'a>(llvm: &LLVM<'a>, index: &mut Index<'a>, data
 /// - an alias index entry for sub-range types
 /// - Array type for arrays
 /// - array type for sized Strings
-pub fn generate_data_type<'a>(
-    module: &Module<'a>,
-    llvm: &LLVM<'a>,
-    index: &mut Index<'a>,
-    data_types: &Vec<UserTypeDeclaration>,
-) -> Result<(), CompileError> {
-    for data_type in data_types {
-        match &data_type.data_type {
-            DataType::StructType { name, variables } => {
-                let name = name.as_ref().unwrap();
-                let mut struct_generator = StructGenerator::new(llvm, index);
-                let members: Vec<&Variable> = variables.iter().collect();
-                let (_, initial_value) = struct_generator.generate_struct_type(&members, name)?;
-                index.associate_type_initial_value(name, initial_value);
-                
-            }
-            DataType::EnumType { name: _, elements } => {
-                // generate a global variable for every enum element
-                for (i, element) in elements.iter().enumerate() {
-                    let int_type = llvm.i32_type();
-                    let element_variable = llvm.create_global_variable(
-                        module, 
-                        element,
-                        int_type.as_basic_type_enum(),
-                        Some(int_type.const_int(i as u64, false).as_basic_value_enum()),
-                    );
+pub fn generate_data_types<'ink, 'ind>(llvm: &LLVM<'ink>, index: &'ind Index) -> Result<LLVMTypedIndex<'ink>, CompileError>{
+    let mut types_index = LLVMTypedIndex::new();
+    let types = index.get_types();
+    for (name, user_type) in types {
+        if let Some(gen_type) =  create_opaque_struct(llvm, user_type) {
```

Comment:
this reads strange, I would suggest we do the struct-check here and only call if we really want an opaque struct.
then the create_opaque_struct becomes a oneliner 

i think its even less code that way

```
if let DataTypeInformation::Struct {name, ...} = user_type {
   types_index.associate_type(name, llvm.create_struct_stub(name).into());
}
```

Reply:
Done